_posts/2021-03-15-elastic-training.md

```diff
@@ -0,0 +1,157 @@
+ 
+1. Fault Tolerance: any worker instance can fail as long as at least one is surviving.
+2. Resources Utilization: when the resources stress piles, the cluster is able to reduce the replicas of workloads with lower priority (distributed training workloads), releasing resource to other workloads (such as prediction service), ensuring SLA for business; after resources released from workloads, elastic training job is able to absorb these resource by scaling up workload replicas.
+3. Training on Cloud: there is a type of resource on the cloud that is called "spot" or "preemptible" instances; it comes with unexptected low price tags but may be retrieved after guaranteed hour expires.
+ 
+Elastic training appears a perfect match to public cloud. Combined with spot instances, we cut the cost for GPUs from ¥16.21/hour to ¥1.62/hour, reducing the overall cost for the training job by nearly 70%. Under the same budget, elastic training employs more GPUs and accelerates the training speed by 5 to 10 times.
+ 
+## Elastic Horovod
+ 
+As the major player in distributed training framework, Horovod v0.20.0 offers its solution to elastic training, [Elastic Horovod](https://horovod.readthedocs.io/en/stable/elastic_include.html). Here we quotes the architecture differences between Elastic Horovod and existing Horovod from [RFC Elastic Horovod](https://docs.google.com/document/d/15ZoHA5AeSI_boeyIBapg9WPXKrYXMRvPytPzQWTCTn4/edit#):
+
+<img src="/images/2021-03-15-elastic-training/horovod-elastic.png" width="" alt="alt_text" title="">
+ 
+* All collective operations are coordinated within a hvd.elastic.run function.
+* State is synchronized between workers before the user’s training function is executed.
+* Worker failure or worker added events will result in a reset event on other workers.
+* Reset events act as barriers to:
+  * Determine whether the job should continue based on worker exit codes.
+  * Blacklist failing hosts.
+  * Launch workers on new hosts.
+  * Update the rank information on existing workers.
+* State is synchronized following a reset event.
+ 
+When launching an elastic training job, `horovodrun` requires a `discover_hosts.sh` script to detect available hosts and slots in real time. (In the following section, we refer this script as `discover_hosts.sh`. Nevertheless the script needs not to be named as `discover_hosts.sh`.)
+ 
+## Elastic Horovod on Kubernetes
+ 
+[MPI-Operator](https://github.com/kubeflow/mpi-operator) is designed to deploy Horovod jobs on Kubernetes. While the operator releases multiple versions, the general idea stays unchanged. It includes:
+
+<img src="/images/2021-03-15-elastic-training/mpi-operator.png" width="" alt="alt_text" title="">
+ 
+1. MPIJob Controller creates a launcher pod and worker pods according to the replicas configuration in MPIJobs
```

Comment:
Should we try to be consistent between graph numeration and these steps ?

Reply:
Since we do not want to re-iterate the mpi-operator proposal in this blog, I changed the numeration to bullet so readers should be not confused. What do you think?

Reply:
Thank you, I think it is better since it will not confuse reader.